File: packages/nexus-reactor/src/parser/lexer.ts
/**
 * @nexus/reactor - NXML Lexer
 * 
 * Tokenizes NXML source into a stream of tokens for the parser.
 */

import type { SourceLocation } from '../core/types';
import { ParseError } from '../core/errors';

// ============================================================================
// Token Types
// ============================================================================

export type TokenType =
  | 'TAG_OPEN'        // <
  | 'TAG_CLOSE'       // >
  | 'TAG_SELF_CLOSE'  // />
  | 'TAG_END_OPEN'    // </
  | 'TAG_NAME'        // NexusPanel, Data, etc.
  | 'ATTR_NAME'       // name, type, etc.
  | 'ATTR_VALUE'      // "value" (includes quotes)
  | 'EQUALS'          // =
  | 'TEXT'            // Text content between tags
  | 'CDATA'           // CDATA content
  | 'COMMENT'         // <!-- comment -->
  | 'EOF';

export interface Token {
  type: TokenType;
  value: string;
  loc: SourceLocation;
}

// ============================================================================
// Lexer State
// ============================================================================

interface LexerState {
  source: string;
  pos: number;
  line: number;
  column: number;
  tokens: Token[];
}

// ============================================================================
// Lexer Implementation
// ============================================================================

/**
 * Tokenize NXML source into tokens
 */
export function tokenize(source: string): Token[] {
  const state: LexerState = {
    source,
    pos: 0,
    line: 1,
    column: 1,
    tokens: [],
  };

  while (!isAtEnd(state)) {
    scanToken(state);
  }

  // Add EOF token
  state.tokens.push({
    type: 'EOF',
    value: '',
    loc: createLocation(state, state.pos),
  });

  return state.tokens;
}

/**
 * Scan the next token
 */
function scanToken(state: LexerState): void {
  skipWhitespace(state);
  
  if (isAtEnd(state)) return;

  const startPos = state.pos;
  const startLine = state.line;
  const startColumn = state.column;

  const char = current(state);

  // Check for comments
  if (char === '<' && lookAhead(state, 1, 3) === '!--') {
    scanComment(state);
    return;
  }

  // Check for CDATA
  if (char === '<' && lookAhead(state, 1, 8) === '![CDATA[') {
    scanCDATA(state);
    return;
  }

  // Check for end tag
  if (char === '<' && peek(state, 1) === '/') {
    advance(state); // <
    advance(state); // /
    addToken(state, 'TAG_END_OPEN', '</', startPos, startLine, startColumn);
    scanTagName(state);
    skipWhitespace(state);
    if (current(state) === '>') {
      const closeStart = state.pos;
      advance(state);
      addToken(state, 'TAG_CLOSE', '>', closeStart, state.line, state.column - 1);
    }
    return;
  }

  // Check for start tag
  if (char === '<') {
    const openStart = state.pos;
    advance(state);
    addToken(state, 'TAG_OPEN', '<', openStart, startLine, startColumn);
    
    skipWhitespace(state);
    scanTagName(state);
    scanAttributes(state);
    
    skipWhitespace(state);
    
    // Check for self-closing or regular close
    if (current(state) === '/' && peek(state, 1) === '>') {
      const selfCloseStart = state.pos;
      advance(state);
      advance(state);
      addToken(state, 'TAG_SELF_CLOSE', '/>', selfCloseStart, state.line, state.column - 2);
    } else if (current(state) === '>') {
      const closeStart = state.pos;
      advance(state);
      addToken(state, 'TAG_CLOSE', '>', closeStart, state.line, state.column - 1);
    } else {
      throw ParseError.unexpectedToken(
        current(state),
        '> or />',
        createLocation(state, state.pos)
      );
    }
    return;
  }

  // Text content
  scanText(state);
}

/**
 * Scan a tag name
 */
function scanTagName(state: LexerState): void {
  skipWhitespace(state);
  
  const startPos = state.pos;
  const startLine = state.line;
  const startColumn = state.column;

  while (!isAtEnd(state) && isNameChar(current(state))) {
    advance(state);
  }

  const value = state.source.slice(startPos, state.pos);
  if (value) {
    addToken(state, 'TAG_NAME', value, startPos, startLine, startColumn);
  }
}

/**
 * Scan attributes
 */
function scanAttributes(state: LexerState): void {
  while (!isAtEnd(state)) {
    skipWhitespace(state);
    
    const char = current(state);
    
    // End of attributes
    if (char === '>' || char === '/') {
      break;
    }

    // Attribute name
    const nameStartPos = state.pos;
    const nameStartLine = state.line;
    const nameStartColumn = state.column;

    while (!isAtEnd(state) && isNameChar(current(state))) {
      advance(state);
    }

    const attrName = state.source.slice(nameStartPos, state.pos);
    if (!attrName) break;

    addToken(state, 'ATTR_NAME', attrName, nameStartPos, nameStartLine, nameStartColumn);

    skipWhitespace(state);

    // Equals sign
    if (current(state) === '=') {
      const eqPos = state.pos;
      advance(state);
      addToken(state, 'EQUALS', '=', eqPos, state.line, state.column - 1);
      
      skipWhitespace(state);

      // Attribute value
      scanAttributeValue(state);
    }
  }
}

/**
 * Scan an attribute value (quoted string)
 */
function scanAttributeValue(state: LexerState): void {
  const startPos = state.pos;
  const startLine = state.line;
  const startColumn = state.column;

  const quote = current(state);
  if (quote !== '"' && quote !== "'") {
    throw ParseError.unexpectedToken(quote, 'quoted string', createLocation(state, state.pos));
  }

  advance(state); // Opening quote

  while (!isAtEnd(state) && current(state) !== quote) {
    if (current(state) === '\\' && peek(state, 1) === quote) {
      advance(state); // Skip escape character
    }
    advance(state);
  }

  if (isAtEnd(state)) {
    throw new ParseError('Unterminated string', { loc: createLocation(state, startPos) });
  }

  advance(state); // Closing quote

  const value = state.source.slice(startPos, state.pos);
  addToken(state, 'ATTR_VALUE', value, startPos, startLine, startColumn);
}

/**
 * Scan text content between tags
 */
function scanText(state: LexerState): void {
  const startPos = state.pos;
  const startLine = state.line;
  const startColumn = state.column;

  while (!isAtEnd(state) && current(state) !== '<') {
    advance(state);
  }

  const value = state.source.slice(startPos, state.pos).trim();
  if (value) {
    addToken(state, 'TEXT', value, startPos, startLine, startColumn);
  }
}

/**
 * Scan a comment: <!-- ... -->
 */
function scanComment(state: LexerState): void {
  const startPos = state.pos;
  const startLine = state.line;
  const startColumn = state.column;

  // Skip <!--
  for (let i = 0; i < 4; i++) advance(state);

  while (!isAtEnd(state)) {
    if (current(state) === '-' && lookAhead(state, 0, 3) === '-->') {
      break;
    }
    advance(state);
  }

  // Skip -->
  for (let i = 0; i < 3 && !isAtEnd(state); i++) advance(state);

  const value = state.source.slice(startPos, state.pos);
  addToken(state, 'COMMENT', value, startPos, startLine, startColumn);
}

/**
 * Scan CDATA: <![CDATA[ ... ]]>
 */
function scanCDATA(state: LexerState): void {
  const startPos = state.pos;
  const startLine = state.line;
  const startColumn = state.column;

  // Skip <![CDATA[
  for (let i = 0; i < 9; i++) advance(state);

  const contentStart = state.pos;

  while (!isAtEnd(state)) {
    if (current(state) === ']' && lookAhead(state, 0, 3) === ']]>') {
      break;
    }
    advance(state);
  }

  const content = state.source.slice(contentStart, state.pos);

  // Skip ]]>
  for (let i = 0; i < 3 && !isAtEnd(state); i++) advance(state);

  addToken(state, 'CDATA', content, startPos, startLine, startColumn);
}

// ============================================================================
// Helper Functions
// ============================================================================

function isAtEnd(state: LexerState): boolean {
  return state.pos >= state.source.length;
}

function current(state: LexerState): string {
  return state.source[state.pos] ?? '';
}

function peek(state: LexerState, offset: number): string {
  return state.source[state.pos + offset] ?? '';
}

function lookAhead(state: LexerState, start: number, length: number): string {
  return state.source.slice(state.pos + start, state.pos + start + length);
}

function advance(state: LexerState): string {
  const char = current(state);
  state.pos++;
  
  if (char === '\n') {
    state.line++;
    state.column = 1;
  } else {
    state.column++;
  }
  
  return char;
}

function skipWhitespace(state: LexerState): void {
  while (!isAtEnd(state) && /\s/.test(current(state))) {
    advance(state);
  }
}

function isNameChar(char: string): boolean {
  return /[a-zA-Z0-9_$-]/.test(char);
}

function createLocation(state: LexerState, startPos: number): SourceLocation {
  // Calculate start line/column from startPos
  let line = 1;
  let column = 1;
  
  for (let i = 0; i < startPos && i < state.source.length; i++) {
    if (state.source[i] === '\n') {
      line++;
      column = 1;
    } else {
      column++;
    }
  }
  
  return {
    startLine: line,
    startColumn: column,
    endLine: state.line,
    endColumn: state.column,
  };
}

function addToken(
  state: LexerState,
  type: TokenType,
  value: string,
  _startPos: number,
  startLine: number,
  startColumn: number
): void {
  state.tokens.push({
    type,
    value,
    loc: {
      startLine,
      startColumn,
      endLine: state.line,
      endColumn: state.column,
    },
  });
}

/**
 * Get token type name for error messages
 */
export function getTokenTypeName(type: TokenType): string {
  const names: Record<TokenType, string> = {
    TAG_OPEN: 'opening tag <',
    TAG_CLOSE: 'closing bracket >',
    TAG_SELF_CLOSE: 'self-closing />',
    TAG_END_OPEN: 'end tag </',
    TAG_NAME: 'tag name',
    ATTR_NAME: 'attribute name',
    ATTR_VALUE: 'attribute value',
    EQUALS: 'equals sign',
    TEXT: 'text content',
    CDATA: 'CDATA section',
    COMMENT: 'comment',
    EOF: 'end of file',
  };
  return names[type] ?? type;
}

File: runtime/nexus-core/nexus_core/parser/lexer.py
"""
NXML Lexer - Tokenizes NXML source code.
"""

from dataclasses import dataclass
from enum import Enum
from typing import List, Optional
import re


class TokenType(Enum):
    """Token types for NXML."""

    # XML structure
    TAG_OPEN = "TAG_OPEN"  # <
    TAG_CLOSE = "TAG_CLOSE"  # >
    TAG_END_OPEN = "TAG_END_OPEN"  # </
    TAG_SELF_CLOSE = "TAG_SELF_CLOSE"  # />
    TAG_NAME = "TAG_NAME"  # Element name
    ATTR_NAME = "ATTR_NAME"  # Attribute name
    ATTR_VALUE = "ATTR_VALUE"  # Attribute value
    EQUALS = "EQUALS"  # =

    # Content
    TEXT = "TEXT"  # Text content
    EXPRESSION = "EXPRESSION"  # {$state.x}
    CODE_BLOCK = "CODE_BLOCK"  # Handler code

    # Special
    EOF = "EOF"
    NEWLINE = "NEWLINE"


@dataclass
class SourceLocation:
    """Source code location for error reporting."""

    start_line: int
    start_column: int
    end_line: int
    end_column: int


@dataclass
class Token:
    """A single token from the lexer."""

    type: TokenType
    value: str
    loc: SourceLocation


class LexerError(Exception):
    """Lexer error with location."""

    def __init__(self, message: str, loc: SourceLocation):
        self.message = message
        self.loc = loc
        super().__init__(f"{message} at line {loc.start_line}, column {loc.start_column}")


class Lexer:
    """NXML lexer with context-aware tokenization."""

    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: List[Token] = []
        self.in_tag = False
        self.in_handler = False

    def tokenize(self) -> List[Token]:
        """Tokenize the entire source."""
        while not self.is_at_end():
            self._scan_token()

        # Add EOF token
        self.tokens.append(
            Token(type=TokenType.EOF, value="", loc=self._current_location())
        )

        return self.tokens

    def _scan_token(self):
        """Scan a single token."""
        char = self.peek()

        # In a tag (parsing attributes)
        if self.in_tag:
            self._skip_whitespace()
            if self.is_at_end():
                return

            char = self.peek()

            # Check for tag close
            if char == "/" and self.peek_next() == ">":
                self.in_tag = False
                self._add_token(TokenType.TAG_SELF_CLOSE, "/>")
                self.advance()
                self.advance()
                return

            if char == ">":
                self.in_tag = False
                self._add_token(TokenType.TAG_CLOSE, ">")
                self.advance()
                return

            # Attribute name
            if char.isalpha() or char == "_":
                self._scan_attribute_name()
                return

            # Equals sign
            if char == "=":
                self._add_token(TokenType.EQUALS, "=")
                self.advance()
                return

            # Attribute value (quoted)
            if char in ('"', "'"):
                self._scan_attribute_value()
                return

            self.advance()
            return

        # Check for opening tag
        if char == "<":
            if self.peek_next() == "/":
                # Closing tag
                self._scan_closing_tag()
                return
            else:
                # Opening tag
                self._scan_opening_tag()
                return

        # Check for expression
        if char == "{" and self.peek_next() == "$":
            self._scan_expression()
            return

        # Check if we're inside a Handler tag (code block)
        if self.in_handler:
            self._scan_code_block()
            return

        # Regular text content
        self._scan_text()

    def _scan_opening_tag(self):
        """Scan opening tag like <NexusPanel>."""
        start_loc = self._current_location()
        self.advance()  # Skip <
        self._add_token(TokenType.TAG_OPEN, "<")

        # Get tag name
        name_start = self.pos
        while not self.is_at_end() and (
            self.peek().isalnum() or self.peek() in ("_", "-", ":")
        ):
            self.advance()

        name = self.source[name_start : self.pos]
        self._add_token(TokenType.TAG_NAME, name)

        # Check if this is a Handler tag (special handling for code blocks)
        if name == "Handler":
            self.in_handler = True

        # Now we're inside the tag (parsing attributes)
        self.in_tag = True

    def _scan_closing_tag(self):
        """Scan closing tag like </NexusPanel>."""
        start_loc = self._current_location()
        self.advance()  # Skip <
        self.advance()  # Skip /
        self._add_token(TokenType.TAG_END_OPEN, "</")

        # Get tag name
        name_start = self.pos
        while not self.is_at_end() and (
            self.peek().isalnum() or self.peek() in ("_", "-", ":")
        ):
            self.advance()

        name = self.source[name_start : self.pos]
        self._add_token(TokenType.TAG_NAME, name)

        # Check if this closes a Handler tag
        if name == "Handler":
            self.in_handler = False

        # Skip whitespace
        self._skip_whitespace()

        # Expect >
        if self.peek() == ">":
            self._add_token(TokenType.TAG_CLOSE, ">")
            self.advance()

    def _scan_attribute_name(self):
        """Scan attribute name."""
        start = self.pos
        while not self.is_at_end() and (
            self.peek().isalnum() or self.peek() in ("_", "-", ":")
        ):
            self.advance()

        name = self.source[start : self.pos]
        self._add_token(TokenType.ATTR_NAME, name)

    def _scan_attribute_value(self):
        """Scan attribute value (quoted string)."""
        quote = self.peek()
        self.advance()  # Skip opening quote

        start = self.pos
        while not self.is_at_end() and self.peek() != quote:
            if self.peek() == "\n":
                self.line += 1
                self.column = 1
            self.advance()

        value = self.source[start : self.pos]
        self._add_token(TokenType.ATTR_VALUE, value)

        if self.peek() == quote:
            self.advance()  # Skip closing quote

    def _scan_text(self):
        """Scan regular text content."""
        start = self.pos
        while not self.is_at_end():
            char = self.peek()

            # Stop at tag or expression
            if char == "<" or (char == "{" and self.peek_next() == "$"):
                break

            if char == "\n":
                self.line += 1
                self.column = 1

            self.advance()

        text = self.source[start : self.pos].strip()
        if text:
            self._add_token(TokenType.TEXT, text)

    def _scan_expression(self):
        """Scan expression like {$state.count}."""
        start = self.pos
        self.advance()  # Skip {

        # Find closing }
        depth = 1
        while not self.is_at_end() and depth > 0:
            if self.peek() == "{":
                depth += 1
            elif self.peek() == "}":
                depth -= 1

            if self.peek() == "\n":
                self.line += 1
                self.column = 1

            self.advance()

        expr = self.source[start : self.pos]
        self._add_token(TokenType.EXPRESSION, expr)

    def _scan_code_block(self):
        """Scan handler code block until </Handler>."""
        start = self.pos

        # Find the closing </Handler> tag
        while not self.is_at_end():
            if (
                self.peek() == "<"
                and self.peek_next() == "/"
                and self.source[self.pos : self.pos + 10] == "</Handler>"
            ):
                break

            if self.peek() == "\n":
                self.line += 1
                self.column = 1

            self.advance()

        code = self.source[start : self.pos].strip()
        if code:
            self._add_token(TokenType.CODE_BLOCK, code)

    def _add_token(self, token_type: TokenType, value: str):
        """Add a token to the list."""
        loc = self._current_location()
        self.tokens.append(Token(type=token_type, value=value, loc=loc))

    def _current_location(self) -> SourceLocation:
        """Get current source location."""
        return SourceLocation(
            start_line=self.line,
            start_column=self.column,
            end_line=self.line,
            end_column=self.column + 1,
        )

    def _skip_whitespace(self):
        """Skip whitespace characters."""
        while not self.is_at_end() and self.peek() in (" ", "\t", "\n", "\r"):
            if self.peek() == "\n":
                self.line += 1
                self.column = 1
            else:
                self.column += 1
            self.pos += 1

    def peek(self) -> str:
        """Peek at current character."""
        if self.is_at_end():
            return "\0"
        return self.source[self.pos]

    def peek_next(self) -> str:
        """Peek at next character."""
        if self.pos + 1 >= len(self.source):
            return "\0"
        return self.source[self.pos + 1]

    def advance(self) -> str:
        """Advance to next character."""
        char = self.source[self.pos]
        self.pos += 1
        self.column += 1
        return char

    def is_at_end(self) -> bool:
        """Check if we've reached the end."""
        return self.pos >= len(self.source)
